core/src/main/scala/core/builder/ServiceSpecValidator.scala

```diff
@@ -442,12 +437,23 @@ case class ServiceSpecValidator(
   private[this] def validateUnionTypeDiscriminatorValuesAreDistinct(): Seq[String] = {
     service.unions.flatMap { union =>
       dupsError(
-        s"Union[${union.name}] discriminator values",
-        union.types.map(getDiscriminatorValue)
+        s"Union[${union.name}] discriminator value",
+        getAllDiscriminatorValues(union),
       )
     }
   }
 
+  private[this] def getAllDiscriminatorValues(union: Union, resolved: Set[String] = Set.empty): Seq[String] = {
+    union.types.flatMap { t =>
+      service.unions.find(_.name == t.`type`) match {
+        case Some(nestedUnion) if !resolved.contains(nestedUnion.name) && union.name != nestedUnion.name => {
+          getAllDiscriminatorValues(nestedUnion, resolved ++ Seq(union.name))
+        }
+        case _ => Seq(getDiscriminatorValue(t))
+      }
+    }
```

Comment:
```suggestion
    if (resolved.contains(union.name)) 
      Seq.empty
    else {
      val resolvedUpdated = resolved + union.name
      val subUnions = union.types.flatMap(t => service.unions.find(_.name == t.`type`))
      union.types.map(getDiscriminatorValue) ++ subUnions.flatMap(su => getAllDiscriminatorValues(su, resolvedUpdated))
    }
```

Reply:
oh I see - I agree this is more complete (though I'm not sure in practice there is any difference). will update

Reply:
Did the added test case make it clearer? The original implementation did not catch the duplicate because it did not add the discriminator value of the union type.

Reply:
yes - though it made me wonder if we should NOT support discriminator values for union type members that are themselves union types as in practice I don't think this will ever work - but for now all good and consistent

Reply:
Yes - totally agree.
I am not sure if we will ever write this discriminator in the json as we want to "flatten" all the discriminators.

In a similar observation, we probably SHOULD validate that all union types from the same family (as in ancestors or children) must all have the same discriminator name.
WDYT of enforcing?

Reply:
```
In a similar observation, we probably SHOULD validate that all union types from the same family (as in ancestors or children) must all have the same discriminator name.
WDYT of enforcing?
```

Good idea - I'll track